### PR TITLE
perf(recipe): add Qwen3.5 text H100 recipe

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do?

Adds the canonical text-only H100 performance recipe for
`Qwen/Qwen3.5-35B-A3B`, focused tests, a verified H100 performance item in the
model verification card, and reusable performance skill/campaign guidance.

This PR is stacked on the public runtime pin in #5449, which pins the current
Megatron-Core integration head published in NVIDIA/Megatron-LM#6370. The
model-card run used ancestor
`606c01046bf411c54e730739844da9e613f84a76`; later commits update tests and
repair only the inactive deterministic torch fallback's FLA 0.4.2 `l2norm`
call. The measured FlashQLA path is unchanged.

## Frozen configuration

- Hardware/precision: 2 nodes, 16 H100s, BF16 compute
- Workload: sequence length 4096, MBS1, GBS1024, 50 optimizer steps
- Parallelism: TP1/PP1/CP1/EP8/ETP1
- Optimizer memory: precision-aware Adam, FP16 main parameters, BF16 gradients
  and moments, no parameter remainders
- GDN: FlashQLA 0.1.2, fused pre-GDR path, fused gated RMSNorm
- MoE: Hopper `torch.grouped_mm`, HybridEP, static capacity 1.05, 64-token
  chunks, 16 dispatcher SMs, 108 preprocessing SMs, and shared-expert overlap
- Execution: eager mode, no activation recompute, no DP/parameter-gather or
  combined EP communication overlap
- Runtime hygiene: rank-local TileLang, Triton, TorchInductor, and HybridEP JIT
  caches, plus a shared Torch extension build cache

The recipe retains the model-native local `aux_loss` coefficient and gradient
clipping. Forced expert balancing and mock data make this a performance
stability benchmark, not convergence evidence.

The TE cross-entropy path is intentional for this mock-data performance
recipe; it does not change the functional-training coverage, which remains
unverified. The card records the Hugging Face repository's top-level
`Qwen3_5MoeForConditionalGeneration` architecture, while the recipe selects
the supported text-only causal-LM provider.

## Reproduce

```bash
git checkout fdc1ac339303330c1a73eaa3d1d9d4bab70c776b
git submodule update --init 3rdparty/Megatron-LM
./scripts/training/train.sh \
  --nodes 2 --gpus-per-node 8 \
  --recipe qwen35_text_35b_a3b_pretrain_16gpu_h100_bf16_config \
  --mode pretrain --max_steps 50 \
  logger.save_config_filepath=work/model-verification/qwen3.5-35b-a3b/h100-performance/resolved-config.yaml
```

The recipe-specific kernel overlay is required and hash-locked in
`examples/model_verification_cards/qwen3.5-35b-a3b/runtime-requirements.txt`.
Install it without replacing the base container's Torch stack before running
the card command:

```bash
uv pip install --system --no-deps --require-hashes \
  -r examples/model_verification_cards/qwen3.5-35b-a3b/runtime-requirements.txt
```

The caller supplies normal site-specific Slurm, container, credential, cache,
and storage wiring outside the card.

## Verified result

A fresh public checkout of Bridge `fdc1ac339303330c1a73eaa3d1d9d4bab70c776b`
and MCore `606c01046bf411c54e730739844da9e613f84a76` completed the exact 50-step
card command. Steps 41-50 averaged **20,966.100 ms / 280.970
model TFLOP/s/GPU**. LM loss was finite from **12.830670** to
**0.008044560**, `skipped=0`, `NaN=0`, the resolved post-setup configuration
was persisted, and every Slurm step exited `0:0`.

The earlier frozen candidate reached 285.18 model TFLOP/s/GPU; the card uses a
portable 280 model TFLOP/s/GPU final-10 gate to tolerate ordinary node-to-node
variance.

The card remains pinned to `fdc1ac339303330c1a73eaa3d1d9d4bab70c776b`,
the exact public commit used by the accepted run. Later equal-input HybridEP
hardening and EOS-only launcher experiments were removed from the final
stack. The current recipe includes only a CI-required static declaration
rewrite for rank-local cache environment variables; its resolved runtime
environment and performance path are equivalent to the measured source, but
the source text is not byte-for-byte identical.

Compared with the previous public 24,561.870 ms / 239.820 TFLOP/s/GPU result,
the frozen candidate reduced step time by 15.90% and increased throughput by
18.91%.

## Optimizations retained

- FlashQLA Hopper GDN backend plus streamed pre-GDR fusion
- fused GDN output gated RMSNorm (about +1.37% in the matched A/B)
- Hopper `torch.grouped_mm` experts with GPU-resident offsets
- HybridEP static capacity and tuned chunk/SM settings
- shared-expert overlap and local auxiliary balancing
- EP8 optimizer sharding plus FP16 precision-aware optimizer main parameters
- isolated compiler/JIT caches and a shared Torch extension build cache for
  deterministic multi-rank startup

Rejected experiments are intentionally not in the frozen recipe: FP8, MBS2,
GDN recompute, CPU offload, combined EP overlap, DP/parameter-gather overlap,
CUDA graphs, global fused residual RMSNorm, NCCL EP, and alternate chunk/SM
settings were slower, unstable, or exceeded memory.

## Public PR stack

- NVIDIA/Megatron-LM#5843: merged GDN/GDN2 refactor base
- NVIDIA/Megatron-LM#5765: GDN2/FLA integration ancestry
- NVIDIA/Megatron-LM#5564: source for the adapted FlashQLA Hopper backend
- NVIDIA/Megatron-LM#5532: source for the adapted streamed pre-GDR fusion
- NVIDIA/Megatron-LM#6370: exact public integration used by this card
- #5449: Bridge MCore pin and hash-locked runtime overlay

The #6370 integration carries adaptations of NVIDIA/Megatron-LM#5982 and
NVIDIA/Megatron-LM#5396. The former is inactive because the accepted recipe
does not recompute; the latter's separate Q/K normalization path is bypassed
by pre-GDR fusion. Neither is credited for the result.

## Validation

- exact 50-step 16xH100 card gate: 280.970 model TFLOP/s/GPU over
  steps 41-50, finite metrics, zero skipped/NaN iterations, exit `0:0`
- focused Qwen3.5 recipe/runtime unit tests
- model-card schema and privacy validation
- Linux `uv lock --check` and all-extras dry-run
- `uv run pre-commit run --all-files`
- independent reproducibility, code, and PR-stack audits

### CODE REVIEW

**Comment by yaoyu-33** [comment]
NVSkills CI infrastructure note for HEAD `956845d1`:

- Initial downstream pipeline `59771891`: `ci-unit-tests` job `375093695` ended with GitLab `failure_reason=scheduler_failure` and an empty trace.
- The missing prerequisite artifacts caused `gate:content:v1` to report `detect-changes artifacts are missing or malformed`; Tier 1/2/3 validation and signature attachment were skipped.
- A clean rerun through `/nvskills-ci` reproduced the same scheduler failure in downstream pipeline `59775448`, job `375106714`.

Therefore the current NVSkills status is an infrastructure failure, not a content-validation finding from this PR. I am not changing skill metadata to guess around it.

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5098/commits/639b3aefb015f79d0bb039a4c96217894884d2d4

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5098/commits/9552de9617e0ae89430f9f80a7ee955442dce262

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/commits/b7e665976

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5098/commits/b7e665976cdd75caa54a175bcd8100be53c849e6

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5098/commits/5c2ea293baa8616c69dbe479d357dec124be954d

**Comment by yaoyu-33** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5098/commits/b3b5c233c08441a45360df5fc878cb2654de9bab
